From eda5fd6b32163c4b8b4254227ddd5cf97203fcc3 Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Mon, 24 Jul 2023 00:12:20 -0700
Subject: [PATCH 1/3] arm64: Disable GENERIC_IRQ_EFFECTIVE_AFF_MASK

The effective affinity mask causes a lot of bugs by virtue of many
set_irq_affinity handlers only setting an effective affinity mask for an
IRQ's parent but not the IRQ itself. Since this is a widespread issue that
would require manual fixing on every different SoC, just disable the
effective affinity mask altogether and use the first CPU in an affinity
mask configured.

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
---
 drivers/irqchip/Kconfig | 2 --
 kernel/irq/Kconfig      | 2 +-
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/drivers/irqchip/Kconfig b/drivers/irqchip/Kconfig
index a29a426e4eed..ae80220f5556 100644
--- a/drivers/irqchip/Kconfig
+++ b/drivers/irqchip/Kconfig
@@ -8,7 +8,6 @@ config IRQCHIP
 config ARM_GIC
 	bool
 	select IRQ_DOMAIN_HIERARCHY
-	select GENERIC_IRQ_EFFECTIVE_AFF_MASK if SMP
 
 config ARM_GIC_PM
 	bool
@@ -34,7 +33,6 @@ config ARM_GIC_V3
 	bool
 	select IRQ_DOMAIN_HIERARCHY
 	select PARTITION_PERCPU
-	select GENERIC_IRQ_EFFECTIVE_AFF_MASK if SMP
 	select HAVE_ARM_SMCCC_DISCOVERY
 
 config ARM_GIC_V3_ITS
diff --git a/kernel/irq/Kconfig b/kernel/irq/Kconfig
index db3d174c53d4..a149bba9d08c 100644
--- a/kernel/irq/Kconfig
+++ b/kernel/irq/Kconfig
@@ -24,7 +24,7 @@ config GENERIC_IRQ_SHOW_LEVEL
 
 # Supports effective affinity mask
 config GENERIC_IRQ_EFFECTIVE_AFF_MASK
-       depends on SMP
+       depends on SMP && !ARM64
        bool
 
 # Support for delayed migration from interrupt context
-- 
2.43.0

